client/protocol.go

```diff
@@ -0,0 +1,413 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	curpapi "github.com/xline-kv/go-xline/api/curp"
+	xlineapi "github.com/xline-kv/go-xline/api/xline"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+)
+
+// Protocol client
+type protocolClient struct {
+	// Curp members
+	addrs []string
+	// All servers's `Connect`
+	connects map[uint64]string
+	// Curp client timeout settings
+	timeout clientTimeout
+}
+
+// Build client from addresses, this method will fetch all members from servers
+func BuildCurpClientFromAddrs(addrs []string) (*protocolClient, error) {
+	clientTimeout := NewDefaultClientTimeout()
+	cluster, err := fetchCluster(addrs, clientTimeout.proposeTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	return &protocolClient{
+		addrs:    addrs,
+		connects: cluster.AllMembers,
+		timeout:  clientTimeout,
+	}, nil
+}
+
+// Fetch cluster from server
+func fetchCluster(addrs []string, proposeTimeout time.Duration) (*curpapi.FetchClusterResponse, error) {
+	var protocolClients []curpapi.ProtocolClient
+	fetchClusterRespCh := make(chan *curpapi.FetchClusterResponse)
+
+	for _, addr := range addrs {
+		conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithIdleTimeout(1*time.Second))
+		if err != nil {
+			logrus.Warn(err)
+		} else {
+			protocolClient := curpapi.NewProtocolClient(conn)
+			protocolClients = append(protocolClients[:], protocolClient)
+		}
+	}
+
+	for _, protocolClient := range protocolClients {
+		client := protocolClient
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), proposeTimeout)
+			defer cancel()
+			res, err := client.FetchCluster(ctx, &curpapi.FetchClusterRequest{})
+			if err != nil {
+				logrus.Debug(err)
+			} else {
+				fetchClusterRespCh <- res
+			}
+		}()
+	}
+
+	select {
+	case fetchClusterRes := <-fetchClusterRespCh:
+		return fetchClusterRes, nil
+	case <-time.After(proposeTimeout):
+		return nil, errors.New("fetch cluster fail")
+	}
```

Comment:
I think we should return an error immediately at the time when all fetch requests are failed. It seems in this code it will still wait for `proposeTimeout`.

Reply:
done